[{"url":"https://api.github.com/repos/coq/coq/issues/comments/494850476","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494850476","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494850476,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDg1MDQ3Ng==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T15:22:27Z","updated_at":"2019-05-22T15:22:27Z","author_association":"CONTRIBUTOR","body":"This is a fun one, if you wrap the Qed in Fail it says the command doesn't fail x_x","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494850476/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494936024","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494936024","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494936024,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDkzNjAyNA==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T19:28:42Z","updated_at":"2019-05-22T19:28:42Z","author_association":"CONTRIBUTOR","body":"So it runs\r\n- let v := ..., producing `v := @id Type@{1} ?bla`\r\n- Axiom admit', producing `admit' : forall T : Type@{3}, T`\r\n- exact admit', inferring universe constraint `1 < 3`\r\n- reset environment to before the proof\r\n- Axiom admit', producing `admit' : forall T : Type@{5}, T`\r\n- save the above proof with `1 < 3`, but 3 doesn't exist\r\n\r\nWith `Fail` it doesn't reset the envionment + rerun Axiom, so universe 3 is still there and saving succeeds.\r\nPossible solutions:\r\n- make Fail run after resetting the environment + rerunning commands? There may be related bugs without using the universe counter because of this difference between Fail and non-Fail\r\n- somehow remove the global, non-resettable universe counter\r\n- some system which doesn't fully rerun commands inside proofs?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494936024/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494945766","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494945766","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494945766,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk0NTc2Ng==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T19:53:29Z","updated_at":"2019-05-22T19:53:29Z","author_association":"MEMBER","body":"In coq-lsp [which just runs vernacinterp linearly] I get an anomaly on the first exact, already. interesting,\r\n\r\n```\r\nConversion test raised an anomaly: Anomaly \"Universe foo.3 undefined.\"\r\n```","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494945766/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494946366","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494946366","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494946366,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk0NjM2Ng==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T19:55:05Z","updated_at":"2019-05-22T19:55:05Z","author_association":"CONTRIBUTOR","body":"Probably missing some `update_global_env` call","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494946366/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494946701","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494946701","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494946701,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk0NjcwMQ==","user":{"login":"maximedenes","id":647105,"node_id":"MDQ6VXNlcjY0NzEwNQ==","avatar_url":"https://avatars.githubusercontent.com/u/647105?v=4","gravatar_id":"","url":"https://api.github.com/users/maximedenes","html_url":"https://github.com/maximedenes","followers_url":"https://api.github.com/users/maximedenes/followers","following_url":"https://api.github.com/users/maximedenes/following{/other_user}","gists_url":"https://api.github.com/users/maximedenes/gists{/gist_id}","starred_url":"https://api.github.com/users/maximedenes/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/maximedenes/subscriptions","organizations_url":"https://api.github.com/users/maximedenes/orgs","repos_url":"https://api.github.com/users/maximedenes/repos","events_url":"https://api.github.com/users/maximedenes/events{/privacy}","received_events_url":"https://api.github.com/users/maximedenes/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T19:55:49Z","updated_at":"2019-05-22T19:55:49Z","author_association":"MEMBER","body":"I thought we had a very similar bug at some point, that was fixed IIRC. What you observe with `Fail` may be due to a different classification.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494946701/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494947462","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494947462","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494947462,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk0NzQ2Mg==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T19:57:49Z","updated_at":"2019-05-22T19:57:49Z","author_association":"MEMBER","body":"> Probably missing some `update_global_env` call\r\n\r\nUh, why is that needed? Does it mean that `Vernacentries.interp` is unsound as it stands?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494947462/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494949548","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494949548","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494949548,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk0OTU0OA==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:03:14Z","updated_at":"2019-05-22T20:03:14Z","author_association":"CONTRIBUTOR","body":"The evar map has a copy of the universe graph, + the proof's local universes.\r\nWhen you do something that declares global universes, such as Axiom, you need to update the evar map with the new universes.\r\n\r\nThis is related to https://github.com/coq/coq/issues/9455","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494949548/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494953312","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494953312","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494953312,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk1MzMxMg==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:13:11Z","updated_at":"2019-05-22T20:13:11Z","author_association":"MEMBER","body":"Yeah so why is not the axiom code updating the corresponding evar maps? After all it has controlled access to the proof stack now.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494953312/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494953625","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494953625","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494953625,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk1MzYyNQ==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:14:03Z","updated_at":"2019-05-22T20:14:03Z","author_association":"CONTRIBUTOR","body":"We don't want random commands to mess with random proofs do we?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494953625/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494954975","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494954975","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494954975,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk1NDk3NQ==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:17:30Z","updated_at":"2019-05-22T20:17:30Z","author_association":"MEMBER","body":"> We don't want random commands to mess with random proofs do we?\r\n\r\nI am not sure what do you mean. Is there any other way?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494954975/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494956797","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494956797","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494956797,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk1Njc5Nw==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:22:20Z","updated_at":"2019-05-22T20:22:20Z","author_association":"CONTRIBUTOR","body":"update_sigma_env before accessing the proof, instead of after modifying the global env","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494956797/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494957465","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494957465","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494957465,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk1NzQ2NQ==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:24:05Z","updated_at":"2019-05-22T20:24:13Z","author_association":"MEMBER","body":"Who should call `update_sigma_env`?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494957465/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494959544","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494959544","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494959544,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk1OTU0NA==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:29:12Z","updated_at":"2019-05-22T20:29:12Z","author_association":"CONTRIBUTOR","body":"Current system works afaict\r\nIf you don't need to worry about the stm (like that ?proof thing) then interp_functional_vernac could do it","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494959544/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494961780","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494961780","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494961780,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk2MTc4MA==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:35:10Z","updated_at":"2019-05-22T20:35:10Z","author_association":"MEMBER","body":"Umm, current system does not work; my LSP server hooks directly on `Vernacentries.interp` and I still get problems. So this is not a STM-specific problem.\r\n\r\nWe have many instances where the lower layers are buggy, then the upper layers [the STM] try to patch up, see for example the discard_proof stuff in the STM. I think I don't need to explain why this is not the best approach.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494961780/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494966662","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494966662","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494966662,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk2NjY2Mg==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:48:00Z","updated_at":"2019-05-22T20:48:00Z","author_association":"MEMBER","body":"> If you don't need to worry about the stm (like that ?proof thing) then interp_functional_vernac could do it\r\n\r\nMaybe I misunderstood the part of \"current system works\" vs \"interp could do it\".\r\n\r\nWhat I was proposing if for interp to do it; but you answered \"We don't want random commands to mess with random proofs\"\r\n\r\nSo I am a bit lost.\r\n\r\nFor now I would be happy if `interp` was sound [forgetting about the `?proof` inject hack] , once we have a sound system we could think about optimizations, but not before.\r\n","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494966662/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494967387","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494967387","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494967387,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk2NzM4Nw==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:50:02Z","updated_at":"2019-05-22T20:50:02Z","author_association":"CONTRIBUTOR","body":"This bug isn't about update_sigma_env so let's stop talking about it here.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494967387/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494969215","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494969215","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494969215,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk2OTIxNQ==","user":{"login":"gares","id":1013846,"node_id":"MDQ6VXNlcjEwMTM4NDY=","avatar_url":"https://avatars.githubusercontent.com/u/1013846?v=4","gravatar_id":"","url":"https://api.github.com/users/gares","html_url":"https://github.com/gares","followers_url":"https://api.github.com/users/gares/followers","following_url":"https://api.github.com/users/gares/following{/other_user}","gists_url":"https://api.github.com/users/gares/gists{/gist_id}","starred_url":"https://api.github.com/users/gares/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/gares/subscriptions","organizations_url":"https://api.github.com/users/gares/orgs","repos_url":"https://api.github.com/users/gares/repos","events_url":"https://api.github.com/users/gares/events{/privacy}","received_events_url":"https://api.github.com/users/gares/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:55:07Z","updated_at":"2019-05-22T20:55:07Z","author_association":"MEMBER","body":"Adding axioms in the middle of a proof is not supported. Can't we all agree on a simple thing like that? It broke badly in 8.5 due to universe polymorphism + STM.\r\n\r\nMy point is: the Coq parser let's you write almost everything. Why should this loose behavior define what is licit? Single instances of this phenomenon may look reasonable, but if you think about it, it lets you require ml code in the middle of a functor signature... The fact that it parses, and hence reaches the interpreter, be it linear or parallel, changes little to me. You can't possibly give a reasonable meaning to any sequence of parsed sentences. The fact that the STM tries to support some of these scenarios is, FTR, not my original intention/design. Was it in any pre 8.5 design? I bet no. \r\n\r\nGoing back to the actual problem... sorry for the digression : Jason, we are all happy to receive minimized bug, but it also helps to understand what you were trying to achieve in your original use case. Why do you need to declare axioms in the middle of something in the first place? There may be, or we could provide, better working ways of achieving your goal.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494969215/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494969440","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494969440","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494969440,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk2OTQ0MA==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:55:45Z","updated_at":"2019-05-22T20:55:45Z","author_association":"MEMBER","body":"> This bug isn't about update_sigma_env so let's stop talking about it here.\r\n\r\nIs it not? You suggested it was due to it, or do you want me to open a different bug for the unsoundness on `interp` ?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494969440/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494970326","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494970326","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494970326,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk3MDMyNg==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:58:10Z","updated_at":"2019-05-22T20:58:10Z","author_association":"MEMBER","body":"> Adding axioms in the middle of a proof is not supported.\r\n\r\nA problem here is that we don't have a proper list of \"known issues\" / \"not supported stuff\"\r\n\r\nI tried to add one in #947 but didn't go so far.\r\n\r\nI do agree that if something is not support then it should fail with an error message instead of producing an anomaly.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494970326/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494970853","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494970853","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494970853,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk3MDg1Mw==","user":{"login":"maximedenes","id":647105,"node_id":"MDQ6VXNlcjY0NzEwNQ==","avatar_url":"https://avatars.githubusercontent.com/u/647105?v=4","gravatar_id":"","url":"https://api.github.com/users/maximedenes","html_url":"https://github.com/maximedenes","followers_url":"https://api.github.com/users/maximedenes/followers","following_url":"https://api.github.com/users/maximedenes/following{/other_user}","gists_url":"https://api.github.com/users/maximedenes/gists{/gist_id}","starred_url":"https://api.github.com/users/maximedenes/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/maximedenes/subscriptions","organizations_url":"https://api.github.com/users/maximedenes/orgs","repos_url":"https://api.github.com/users/maximedenes/repos","events_url":"https://api.github.com/users/maximedenes/events{/privacy}","received_events_url":"https://api.github.com/users/maximedenes/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T20:59:33Z","updated_at":"2019-05-22T20:59:33Z","author_association":"MEMBER","body":"> A problem here is that we don't have a proper list of \"known issues\" / \"not supported stuff\"\r\n\r\nMaybe it would be easier to define what is supported, i.e. a document grammar?","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494970853/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494971064","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494971064","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494971064,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk3MTA2NA==","user":{"login":"gares","id":1013846,"node_id":"MDQ6VXNlcjEwMTM4NDY=","avatar_url":"https://avatars.githubusercontent.com/u/1013846?v=4","gravatar_id":"","url":"https://api.github.com/users/gares","html_url":"https://github.com/gares","followers_url":"https://api.github.com/users/gares/followers","following_url":"https://api.github.com/users/gares/following{/other_user}","gists_url":"https://api.github.com/users/gares/gists{/gist_id}","starred_url":"https://api.github.com/users/gares/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/gares/subscriptions","organizations_url":"https://api.github.com/users/gares/orgs","repos_url":"https://api.github.com/users/gares/repos","events_url":"https://api.github.com/users/gares/events{/privacy}","received_events_url":"https://api.github.com/users/gares/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T21:00:10Z","updated_at":"2019-05-22T21:00:10Z","author_association":"MEMBER","body":"> I do agree that if something is not support then it should fail with an error message instead of producing an anomaly.\r\n\r\nI fully agree. And I think the most civilized way is a parse error.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494971064/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/494971583","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-494971583","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":494971583,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NDk3MTU4Mw==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-22T21:01:38Z","updated_at":"2019-05-22T21:01:38Z","author_association":"MEMBER","body":"> Adding axioms in the middle of a proof is not supported. Can't we all agree on a simple thing like that?\r\n\r\nWell indeed that may be tricky for example for the same reason users complained about the \"deprecation\" of nested proofs.\r\n\r\nI am unfamiliar with the technical difficulties of supporting axioms inside a proof; why is it hard? ","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/494971583/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/495271219","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-495271219","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":495271219,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NTI3MTIxOQ==","user":{"login":"JasonGross","id":396076,"node_id":"MDQ6VXNlcjM5NjA3Ng==","avatar_url":"https://avatars.githubusercontent.com/u/396076?v=4","gravatar_id":"","url":"https://api.github.com/users/JasonGross","html_url":"https://github.com/JasonGross","followers_url":"https://api.github.com/users/JasonGross/followers","following_url":"https://api.github.com/users/JasonGross/following{/other_user}","gists_url":"https://api.github.com/users/JasonGross/gists{/gist_id}","starred_url":"https://api.github.com/users/JasonGross/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/JasonGross/subscriptions","organizations_url":"https://api.github.com/users/JasonGross/orgs","repos_url":"https://api.github.com/users/JasonGross/repos","events_url":"https://api.github.com/users/JasonGross/events{/privacy}","received_events_url":"https://api.github.com/users/JasonGross/received_events","type":"User","site_admin":false},"created_at":"2019-05-23T15:37:28Z","updated_at":"2019-05-23T15:37:28Z","author_association":"MEMBER","body":"> Jason, we are all happy to receive minimized bug, but it also helps to understand what you were trying to achieve in your original use case. Why do you need to declare axioms in the middle of something in the first place? \r\n\r\nWhen I have slow proof scripts, I sometimes want to prototype things by adding axioms without having to re-run the entire script.  I reported this because it's an anomaly; if you make it an error to close a proof with an axiom in the middle of it (as long as the error shows up at Qed-time and there is a warning and not an error at Axiom declaration time), I will not complain.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/495271219/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/495272843","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-495272843","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":495272843,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NTI3Mjg0Mw==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-23T15:41:22Z","updated_at":"2019-05-23T15:41:22Z","author_association":"MEMBER","body":"> if you make it an error to close a proof with an axiom in the middle of it (as long as the error shows up at Qed-time and there is a warning and not an error at Axiom declaration time), I will not complain.\r\n\r\nUnfortunately @JasonGross this bug happens before Qed time too.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/495272843/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/495807488","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-495807488","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":495807488,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NTgwNzQ4OA==","user":{"login":"gares","id":1013846,"node_id":"MDQ6VXNlcjEwMTM4NDY=","avatar_url":"https://avatars.githubusercontent.com/u/1013846?v=4","gravatar_id":"","url":"https://api.github.com/users/gares","html_url":"https://github.com/gares","followers_url":"https://api.github.com/users/gares/followers","following_url":"https://api.github.com/users/gares/following{/other_user}","gists_url":"https://api.github.com/users/gares/gists{/gist_id}","starred_url":"https://api.github.com/users/gares/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/gares/subscriptions","organizations_url":"https://api.github.com/users/gares/orgs","repos_url":"https://api.github.com/users/gares/repos","events_url":"https://api.github.com/users/gares/events{/privacy}","received_events_url":"https://api.github.com/users/gares/received_events","type":"User","site_admin":false},"created_at":"2019-05-24T22:48:02Z","updated_at":"2019-05-24T22:48:02Z","author_association":"MEMBER","body":"> I am unfamiliar with the technical difficulties of supporting axioms inside a proof; why is it hard?\r\n\r\nThe bug we see here is that the universes occurring in the type of the axiom have not only to be added to the global environment, but also the ongoing proof... and given that there is an anomaly, to some other piece of state...\r\n\r\nThe difficulty is that he only semantics we could give to that nested Axiom (or nested proof) is as if the Axiom was declared before the proof was started, but time traveling is still hard in 2019 ;-) It is a scope extrusion, that does work bad in imperative programming and does not work at all in functional programming. At least, this is my understanding of this class of bugs.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/495807488/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/495810440","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-495810440","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":495810440,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NTgxMDQ0MA==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-05-24T23:06:56Z","updated_at":"2019-05-24T23:06:56Z","author_association":"CONTRIBUTOR","body":">The bug we see here is that the universes occurring in the type of the axiom have not only to be added to the global environment, but also the ongoing proof... and given that there is an anomaly, to some other piece of state...\r\n\r\nThat's the lsp bug, not the original bug.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/495810440/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/495957981","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-495957981","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":495957981,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NTk1Nzk4MQ==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-26T00:31:02Z","updated_at":"2019-05-26T00:31:02Z","author_association":"MEMBER","body":"> The difficulty is that he only semantics we could give to that nested Axiom (or nested proof) is as if the Axiom was declared before the proof was started, but time traveling is still hard in 2019 ;-) It is a scope extrusion, that does work bad in imperative programming and does not work at all in functional programming. At least, this is my understanding of this class of bugs.\r\n\r\nThanks for the explanation @gares , but I am not sure how this cannot work on a functional style. Why is not updating the proof stack after adding the axiom enough?\r\n\r\nYou may need some scoping on universes for this to work well tho.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/495957981/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/496029035","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-496029035","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":496029035,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NjAyOTAzNQ==","user":{"login":"gares","id":1013846,"node_id":"MDQ6VXNlcjEwMTM4NDY=","avatar_url":"https://avatars.githubusercontent.com/u/1013846?v=4","gravatar_id":"","url":"https://api.github.com/users/gares","html_url":"https://github.com/gares","followers_url":"https://api.github.com/users/gares/followers","following_url":"https://api.github.com/users/gares/following{/other_user}","gists_url":"https://api.github.com/users/gares/gists{/gist_id}","starred_url":"https://api.github.com/users/gares/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/gares/subscriptions","organizations_url":"https://api.github.com/users/gares/orgs","repos_url":"https://api.github.com/users/gares/repos","events_url":"https://api.github.com/users/gares/events{/privacy}","received_events_url":"https://api.github.com/users/gares/received_events","type":"User","site_admin":false},"created_at":"2019-05-26T20:27:14Z","updated_at":"2019-05-26T20:29:21Z","author_association":"MEMBER","body":"> updating the proof stack after adding the axiom enough?\r\n\r\nIf your entire state is reified, then you can surely traverse it and patch it. Any piece of data that was synthesized looking at the data that changed has to be recomputed. One is the set of existing universes inside the evar_map: you have to patch all its copies referenced by your state, and all the data derived from it. Some other piece of state may not be reified, eg a hook to be called at proof closing (the outdated data may be in the closure).\r\n\r\nOf course one can roll back, a re-execute things in the right order. This is sound, but from the user perspective it is as efficient as backtracking the script and placing the axiom were it belongs, that is before the proof starts. So this solution is not good either, or better it is as good as not supporting nested axioms at all, that is my preferred option.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/496029035/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/496035056","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-496035056","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":496035056,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5NjAzNTA1Ng==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-05-26T22:02:53Z","updated_at":"2019-05-26T22:02:53Z","author_association":"MEMBER","body":"> hook to be called at proof closing (the outdated data may be in the closure).\r\n\r\nInteresting thanks, proof closing hooks and data in closures are gone in the PRs reflected in #10041 . Still there is quite a bit work but a lot more control over proof closing is possible after them.\r\n\r\n","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/496035056/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/498063766","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-498063766","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":498063766,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5ODA2Mzc2Ng==","user":{"login":"SkySkimmer","id":2461932,"node_id":"MDQ6VXNlcjI0NjE5MzI=","avatar_url":"https://avatars.githubusercontent.com/u/2461932?v=4","gravatar_id":"","url":"https://api.github.com/users/SkySkimmer","html_url":"https://github.com/SkySkimmer","followers_url":"https://api.github.com/users/SkySkimmer/followers","following_url":"https://api.github.com/users/SkySkimmer/following{/other_user}","gists_url":"https://api.github.com/users/SkySkimmer/gists{/gist_id}","starred_url":"https://api.github.com/users/SkySkimmer/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/SkySkimmer/subscriptions","organizations_url":"https://api.github.com/users/SkySkimmer/orgs","repos_url":"https://api.github.com/users/SkySkimmer/repos","events_url":"https://api.github.com/users/SkySkimmer/events{/privacy}","received_events_url":"https://api.github.com/users/SkySkimmer/received_events","type":"User","site_admin":false},"created_at":"2019-06-02T20:42:17Z","updated_at":"2019-06-02T20:42:17Z","author_association":"CONTRIBUTOR","body":"This is basically the second part of #4603 (it's really 2 bugs in one) which was \"fixed\" in 927e88de465e94bf18b970b45faaf97b4c2dff82 by adding a special case for `Definition := ...`. Since here was with an axiom instead of a definition we see the limitations of this approach...","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/498063766/reactions","total_count":1,"+1":1,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/498066070","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-498066070","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":498066070,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5ODA2NjA3MA==","user":{"login":"ejgallego","id":7192257,"node_id":"MDQ6VXNlcjcxOTIyNTc=","avatar_url":"https://avatars.githubusercontent.com/u/7192257?v=4","gravatar_id":"","url":"https://api.github.com/users/ejgallego","html_url":"https://github.com/ejgallego","followers_url":"https://api.github.com/users/ejgallego/followers","following_url":"https://api.github.com/users/ejgallego/following{/other_user}","gists_url":"https://api.github.com/users/ejgallego/gists{/gist_id}","starred_url":"https://api.github.com/users/ejgallego/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/ejgallego/subscriptions","organizations_url":"https://api.github.com/users/ejgallego/orgs","repos_url":"https://api.github.com/users/ejgallego/repos","events_url":"https://api.github.com/users/ejgallego/events{/privacy}","received_events_url":"https://api.github.com/users/ejgallego/received_events","type":"User","site_admin":false},"created_at":"2019-06-02T21:14:40Z","updated_at":"2019-06-02T21:14:40Z","author_association":"MEMBER","body":"Good catch @SkySkimmer ","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/498066070/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/498122928","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-498122928","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":498122928,"node_id":"MDEyOklzc3VlQ29tbWVudDQ5ODEyMjkyOA==","user":{"login":"gares","id":1013846,"node_id":"MDQ6VXNlcjEwMTM4NDY=","avatar_url":"https://avatars.githubusercontent.com/u/1013846?v=4","gravatar_id":"","url":"https://api.github.com/users/gares","html_url":"https://github.com/gares","followers_url":"https://api.github.com/users/gares/followers","following_url":"https://api.github.com/users/gares/following{/other_user}","gists_url":"https://api.github.com/users/gares/gists{/gist_id}","starred_url":"https://api.github.com/users/gares/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/gares/subscriptions","organizations_url":"https://api.github.com/users/gares/orgs","repos_url":"https://api.github.com/users/gares/repos","events_url":"https://api.github.com/users/gares/events{/privacy}","received_events_url":"https://api.github.com/users/gares/received_events","type":"User","site_admin":false},"created_at":"2019-06-03T05:58:39Z","updated_at":"2019-06-03T05:58:39Z","author_association":"MEMBER","body":"Note that any command with a side effect in the middle of a proof may need special treatment... So I'd say I can't possibly work to allow anything everywhere.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/498122928/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"performed_via_github_app":null},{"url":"https://api.github.com/repos/coq/coq/issues/comments/862514262","html_url":"https://github.com/coq/coq/issues/9912#issuecomment-862514262","issue_url":"https://api.github.com/repos/coq/coq/issues/9912","id":862514262,"node_id":"MDEyOklzc3VlQ29tbWVudDg2MjUxNDI2Mg==","user":{"login":"Columbus240","id":8899730,"node_id":"MDQ6VXNlcjg4OTk3MzA=","avatar_url":"https://avatars.githubusercontent.com/u/8899730?v=4","gravatar_id":"","url":"https://api.github.com/users/Columbus240","html_url":"https://github.com/Columbus240","followers_url":"https://api.github.com/users/Columbus240/followers","following_url":"https://api.github.com/users/Columbus240/following{/other_user}","gists_url":"https://api.github.com/users/Columbus240/gists{/gist_id}","starred_url":"https://api.github.com/users/Columbus240/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/Columbus240/subscriptions","organizations_url":"https://api.github.com/users/Columbus240/orgs","repos_url":"https://api.github.com/users/Columbus240/repos","events_url":"https://api.github.com/users/Columbus240/events{/privacy}","received_events_url":"https://api.github.com/users/Columbus240/received_events","type":"User","site_admin":false},"created_at":"2021-06-16T16:13:26Z","updated_at":"2021-06-16T16:13:26Z","author_association":"CONTRIBUTOR","body":"The problem still occurs with versions 8.13.2 and the current master branch, i.e. commit a1fac38181. Interestingly, if universe polymorphism is enabled, the anomaly doesn’t arise. The problem might be easier to solve, if https://github.com/coq/ceps/pull/42 is implemented.\r\n\r\nRegarding\r\n> When I have slow proof scripts, I sometimes want to prototype things by adding axioms without having to re-run the entire script. I reported this because it's an anomaly; if you make it an error to close a proof with an axiom in the middle of it (as long as the error shows up at Qed-time and there is a warning and not an error at Axiom declaration time), I will not complain.\r\n\r\nIf declaring axioms in the middle of a proof were forbidden, you could work around this restriction using `assert` and `admit`.\r\nReplacing `Axiom Foo : Bar.` with `assert Bar as Foo. { admit. }`.","reactions":{"url":"https://api.github.com/repos/coq/coq/issues/comments/862514262/reactions","total_count":2,"+1":1,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":1,"rocket":0,"eyes":0},"performed_via_github_app":null}]